onadata/apps/viewer/models/data_dictionary.py

```diff
@@ -269,3 +272,142 @@ def save_project(sender, instance=None, created=False, **kwargs):
 pre_save.connect(
     save_project, sender=DataDictionary, dispatch_uid="save_project_datadictionary"
 )
+
+
+def create_registration_form(sender, instance=None, created=False, **kwargs):
+    """Create a RegistrationForm for a form that defines entities
+
+    Create an EntityList if it does not exist. If it exists, use the
+    the existing EntityList
+    """
+    instance_json = instance.json
+
+    if isinstance(instance_json, str):
+        instance_json = json.loads(instance_json)
+
+    if not instance_json.get("entity_related"):
+        return
+
+    children = instance_json.get("children", [])
+    meta_list = filter(lambda child: child.get("name") == "meta", children)
+
+    for meta in meta_list:
+        for child in meta.get("children", []):
+            if child.get("name") == "entity":
+                parameters = child.get("parameters", {})
+                dataset = parameters.get("dataset")
+                entity_list, _ = EntityList.objects.get_or_create(
+                    name=dataset, project=instance.project
+                )
+                (
+                    registration_form,
+                    registration_form_created,
+                ) = RegistrationForm.objects.get_or_create(
+                    entity_list=entity_list,
+                    xform=instance,
+                )
+
+                if registration_form_created:
+                    # RegistrationForm contributing to any previous
+                    # EntityList should be disabled
+                    for form in instance.registration_forms.exclude(
+                        entity_list=entity_list, is_active=True
+                    ):
+                        form.is_active = False
+                        form.save()
+                else:
+                    # If previously disabled, enable it
+                    registration_form.is_active = True
+                    registration_form.save()
+
+                return
+
+
+post_save.connect(
+    create_registration_form,
+    sender=DataDictionary,
+    dispatch_uid="create_registration_form_datadictionary",
+)
+
```

Comment:
I am thinking at some future point, we may want to not have the `DataDictionary`. The earlier thinking of this class was to drive the UI hence it being under the `viewer` module. These functionalities seem tied to the `XForm` object and can/is independent of the views we build. Perhaps we move these functions under the `onadata.logger.models.xform` module.

Reply:
@ukanga The issue after testing while trying to implement the above change is that the signals sent are for the `DataDictionary` model and no signals are sent for the `XForm` model when an a form is published. I think at this point we are stuck with using the `DataDictionary` until we decide to remove it entirely

Reply:
Well noted. Let's stick with it for now.